zcash_client_backend/src/zip321.rs

```diff
@@ -0,0 +1,1046 @@
+                .map(|s| s.map(Address::Sapling))
+                .map_err(AddrParseError::SAddrError);
+
+            match t_res {
+                Err(_) | Ok(None) => s_res,
+                t_addr => t_addr,
+            }
+        }
+    }
+}
+
+pub struct RawMemo([u8; 512]);
+
+#[derive(Debug)]
+pub enum MemoError {
+    InvalidBase64(base64::DecodeError),
+    LengthExceeded(usize),
+}
+
+impl RawMemo {
+    // Construct a memo from a vector of bytes.
+    pub fn from_str(s: &str) -> Result<Self, MemoError> {
+        RawMemo::from_bytes(s.as_bytes())
+    }
+
+    pub fn from_bytes(v: &[u8]) -> Result<Self, MemoError> {
+        if v.len() > 512 {
+            Err(MemoError::LengthExceeded(v.len()))
+        } else {
+            let mut memo: [u8; 512] = [0; 512];
+            memo[..v.len()].copy_from_slice(&v);
+            Ok(RawMemo(memo))
+        }
+    }
+
+    pub fn to_base64(&self) -> String {
+        // strip trailing zero bytes.
+        let mut last_nonzero = -1;
+        for i in (0..(self.0.len())).rev() {
+            if self.0[i] != 0x0 {
+                last_nonzero = i as i64;
+                break;
+            }
+        }
+
+        base64::encode_config(
+            &self.0[..((last_nonzero + 1) as usize)],
+            base64::URL_SAFE_NO_PAD,
+        )
+    }
+
+    pub fn from_base64(s: &str) -> Result<Self, MemoError> {
+        base64::decode_config(s, base64::URL_SAFE_NO_PAD)
+            .map_err(MemoError::InvalidBase64)
+            .and_then(|b| RawMemo::from_bytes(&b))
+    }
+}
+
+impl Debug for RawMemo {
+    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        f.debug_struct("RawMemo")
+            .field("memo", &format!("{:?}...", &self.0[0..17]))
+            .finish()
+    }
+}
+
+impl PartialEq for RawMemo {
+    fn eq(&self, other: &Self) -> bool {
+        self.to_base64() == other.to_base64()
+    }
+}
+
+impl FromStr for RawMemo {
+    type Err = MemoError;
+
+    fn from_str(memo: &str) -> Result<Self, Self::Err> {
+        RawMemo::from_str(memo)
+    }
+}
+
+impl Eq for RawMemo {}
+
+impl PartialOrd for RawMemo {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.to_base64().cmp(&other.to_base64()))
+    }
+}
+
+impl Ord for RawMemo {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.partial_cmp(other).unwrap()
+    }
+}
+
+// RawMemo is somewhat duplicative of the `Memo` type
+// in crate::note_encryption but as that's actively being
+// updated at time of this writing, these functions provide
+// shims to ease future use of those
+pub fn memo_from_vec(v: &[u8]) -> Result<RawMemo, MemoError> {
+    RawMemo::from_bytes(v)
+}
+
+pub fn memo_to_base64(memo: &RawMemo) -> String {
+    memo.to_base64()
+}
+
+pub fn memo_from_base64(s: &str) -> Result<RawMemo, MemoError> {
+    RawMemo::from_base64(s)
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Zip321Payment {
```

Comment:
The naming of this struct confuses me. Isn't the intent of a ZIP 321 URI to request a single payment to potentially multiple recipients? So this would be something like `Recipient` or `RequestedRecipient`.

Separately, we should [avoid redundant prefixes](https://doc.rust-lang.org/1.0.0/style/style/naming/README.html#avoid-redundant-prefixes-[rfc-356]); these can be referred to as `zip321::PaymentRequest` etc. anywhere the prefix matters.

Reply:
We define `Payment` in ZIP-321 as "A transfer of funds implemented by a shielded or transparent output of a single Zcash transaction." We needed a word for that, but perhaps there's a better one. "recipient" doesn't quite capture it, I don't think. Anyway this is consistent with the ZIP.